fix: wrongly calculated amount if inflow and outflow are strings and inflow is 0

tobim-dev opened the pull request:
Given that the `inflow` and `outflow` columns are used, both of which are strings, and the inflow column is `'0'`, the calculated amount is always `0`. This is because JavaScript takes the left value of an II expression if it's truthy, and `'0'` is considered truthy.

In the fix, I explicitly check for `'0'`.

nielsmaerten commented:
Hi Tobias, thanks for discovering this bug and writing a fix!

It looks like this fix is causing one of the [existing tests](https://github.com/nielsmaerten/ynab-buddy/blob/c15ff341e69b41e74b366289d8eae233d43a18af/src/lib/parser.spec.ts#L174) to break though. Could you maybe take another look?
Also before I merge a change, I like to make sure there's an additional test to demonstrate the original bug that the PR is going to fix.

Here's a quick draft of what that might look like:

```diff
diff --git a/src/lib/parser.spec.fixtures.ts b/src/lib/parser.spec.fixtures.ts
index 2fe4b98..f63e647 100644
--- a/src/lib/parser.spec.fixtures.ts
+++ b/src/lib/parser.spec.fixtures.ts
@@ -67,6 +67,15 @@ invalid footer row
 ,,30/08/2020,POS-xxxxxxx,,-7.00,EUR
 ,01/09/2020,01/09/2020,Disposizione - RIF:xxxxxxx,,-180.00,EUR
 ,01/09/2020,31/08/2020,Addebito canone -xxxxxxx,100.00,,EUR`,
+
+  // https://github.com/nielsmaerten/ynab-buddy/pull/269
+  // This demonstrates a bug discovered by @tobim-dev:
+  // If there's an inflow column containing "0", 
+  // the parser should ignore it if there's also an outflow column.
+  // Use case: bank uses inflow/outflow columns,
+  // and in case of outflow writes "0" in inflow column:
+  `index,date,inflow,outflow,payee
+  1,9/27/2020,0,420.69,Devpoint`
 ];
 
 export const defaultParser: Parser =
diff --git a/src/lib/parser.spec.ts b/src/lib/parser.spec.ts
index e8b341a..1f868a8 100644
--- a/src/lib/parser.spec.ts
+++ b/src/lib/parser.spec.ts
@@ -189,6 +189,17 @@ describe("parser", () => {
     expect(result.transactions[1].amount).toEqual(-180);
     expect(result.transactions[2].amount).toEqual(100);
   });
+
+  it("handles an inflow column containing '0'", () => {
+    const parseCfg = {
+      columns: ["", "date", "inflow", "outflow", "payee"],
+      decimal_separator: ".",
+      delimiter: ",",
+      header_rows: 1,
+    };
+    const result = runParser(csvFixtures.inflow0, parseCfg);
+    expect(result.transactions[0].amount).toEqual(-420.69);
+  });
 });
 
 const runParser = (fixtureId: number, parseCfg?: Partial<Parser>) => {
@@ -211,6 +222,7 @@ enum csvFixtures {
   dotThousandSeparatorsCommaDecimalSeparator = 8,
   inflowOutflowColumns = 9,
   githubIssue45 = 10,
+  inflow0 = 11,
 }
 
 const validateTransaction = (tx: Transaction) => {
```

Thanks again!

tobim-dev commented:
Thank you for your review and for reminding me to be more rigorous. :)

Thanks as well for the test case draft; it saved me some effort! I've added it.

I also looked into the failing test case. The initial check for a truthy inflow or outflow value was correct. Checking only for `undefined` is insufficient, so I changed it to check for a truthy inflow or outflow.

I hope this meets all the prerequisites for merging the PR.

nielsmaerten commented:
@tobim-dev Merged, thanks! 🚀

I was looking at the `getValue()` function, and I think we can make it even shorter.  
Then we can also re-use it later in the file, when we need to detect if we're dealing with an "outgoing" transaction.

I put the change in #270, curious to hear what you think :)
